### PR TITLE
Try: building blocks from a npm package

### PR DESCRIPTION
**Pull request by ryelle**
Closes https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/416, see https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/456

This PR adds the [@woocommerce/block-library](https://www.npmjs.com/package/@woocommerce/block-library) package as an dependency to package.json, and updates the grunt build process to pull in the compiled files from the npm package – the CSS and JS files, along with 2 PHP files. One php file, `class-wc-block-library.php` registers the blocks, and `class-wc-block-featured-product.php` handles rendering the Featured Product block.

### How to test the changes in this Pull Request:

There are two parts to this PR; the part where the blocks have been added to the repo, and the part where we actually build the blocks from the block-library package.

To test the first part:

1. Deactivate the blocks plugin if you have it
1. Load up the block editor on a post or a page
2. Add a new block by clicking the +
3. You should see a new section, "WooCommerce", with 8 blocks, including "Products by Category", "Hand-picked Products", "Newest Products", etc.
4. Adding one of these blocks works as expected
5. Publish & view the post
6. The post on the frontend looks the same as the preview

To test the build process:

1. Run `npm install`
2. Run `grunt blocks` or just `grunt` ¹
3. You'll see 3 "copy" tasks run, each copying over some files:

```
Running "copy:js" (copy) task
Copied 10 files

Running "copy:css" (copy) task
Copied 3 files

Running "copy:php" (copy) task
Copied 2 files
```

After copying the files, you can run through the steps in the first part to test that the blocks are working correctly.

¹ I've noticed the grunt task doesn't complete without errors, so I'm not sure if I need to fix anything- the issues i see are in jshint & postcss.

### CODE REVIEW

**Comment by coderkevin** [comment]
Just tried this out, and I'm getting errors for `wcSettings`:

```
Uncaught ReferenceError: wcSettings is not defined
    at Object.window.webpackJsonp (vendors.js:19)
    at i (handpicked-products.js:1)
    at Module.<anonymous> (handpicked-products.js:69)
    at i (handpicked-products.js:1)
    at n (handpicked-products.js:1)
    at handpicked-products.js:1
    at handpicked-products.js:1
```

I'll look further into this, but I figured I'd post my initial result.

**Comment by ryelle** [comment]
Ah, should have mentioned that - yeah that's expected, i don't think it actually breaks the blocks though. I wanted to look into another method for setting the wcSettings global that might be more stable than "copy from wc-admin"

**Comment by coderkevin** [comment]
Hmm, none of the blocks are loading for me and I thought this was why. I even checked `wp.blocks.getBlockTypes()` from the JS console and I only see the 70 core blocks there.

**Comment by ryelle** [comment]
I've updated the branch on the blocks plugin to add wcSettings back in, if you grab that and run `grunt blocks` it should fix the files in woocommerce too.

**Comment by timmyc** [comment]
Also thinking we need to get this PR landed in the feature branch in order for all things to work:

https://github.com/woocommerce/woocommerce/pull/22856

**Comment by coderkevin** [comment]
With the last update, I get the blocks, but 404s from the API calls, so yeah, looks like we will need #22856

**Comment by ryelle** [comment]
Updated with a rebase on the branch in #22856, since it looks like that PR might be held up a little.

**Comment by ryelle** [comment]
Tagging in @woocommerce/proton to get some feedback on this approach - you'll also want to look at https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/456, which has more details on the feature plugin side of things. Let me know if anything's unclear, happy to chat this through in slack too.

File: includes/blocks/class-wc-block-featured-product.php [hunk 1]
```diff
@@ -0,0 +1,176 @@
+<?php
+/**
+ * Display the featured product block in the content.
+ * NOTE: DO NOT edit this file in WooCommerce core, this is generated from woocommerce-gutenberg-products-block.
```
Comment (timmyc) [inline includes/blocks/class-wc-block-featured-product.php]: 👍

**Comment by timmyc** [comment]
FYI I have updated #22856 per the feedback, though will need to do a PR in the blocks repo due to the changes there

**Review by timmyc — COMMENTED** [review]
I was able to test out the first part of the PR just fine, all blocks are working for me as expected 🙌 - I ran into some issues running grunt locally. Code looks good to me and seems like this will be an easy build process.

File: includes/blocks/class-wc-block-library.php [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+		<script type="text/javascript">
+			<?php // Use the wcSettings from wc-admin if already present. ?>
+			var wcSettings = wcSettings || <?php echo wp_json_encode( $settings ); ?>;
+			var wc_product_block_data = <?php echo wp_json_encode( $block_settings ); ?>;
+		</script>
+		<?php
+	}
+}
+
+if ( function_exists( 'register_block_type' ) ) {
```
Comment (claudiosanches) [inline includes/blocks/class-wc-block-library.php]: Can we move this into a `init` method?
So will make easy to make it a singleton or something like when we start implement https://github.com/woocommerce/woocommerce/pull/22849

**Comment by claudiosanches** [comment]
In the editor looks fine, but I had a trouble with the build.

Getting this error:

```
$ npm install
npm ERR! path /***/wp-content/plugins/woocommerce/node_modules/.staging/@woocommerce/block-library-9f746df2/node_modules/@babel/code-frame
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall rename
npm ERR! enoent ENOENT: no such file or directory, rename '/***/wp-content/plugins/woocommerce/node_modules/.staging/@woocommerce/block-library-9f746df2/node_modules/@babel/code-frame' -> '/***/wp-content/plugins/woocommerce/node_modules/.staging/@babel/code-frame-57dfb60c'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! A complete log of this run can be found in:
npm ERR!     /***/.npm/_logs/2019-02-26T23_46_22_159Z-debug.log
```

Congrats in the concept on the build!

File: includes/blocks/class-wc-block-library.php (hunk 2 above)
Comment (ryelle) [reply includes/blocks/class-wc-block-library.php]: If we do that, will we still be able to remove the hooks [with `remove_action` in the feature plugin?](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/5daae873583573968ece147588f91ef860d793da/woocommerce-gutenberg-products-block.php#L29-L32) I started with a singleton implementation but was having trouble getting that to work (which is needed so we can override each hook in the feature plugin)

**Comment by ryelle** [comment]
@claudiosanches @timmyc I was able to replicate the issue you're having, and the easiest fix seems to be to loading from a packaged zip file instead of a local folder. I think `npm link` should also work, but this seemed to be the fastest way to get this testable.

The zip file is hosted on my server, and was created by `npm pack`, which is the same process that creates the package you download from npm.

File: includes/blocks/class-wc-block-library.php (hunk 2 above)
Comment (claudiosanches) [reply includes/blocks/class-wc-block-library.php]: Since it's all static methods, should be fine to remove any action later.

**Comment by timmyc** [comment]
With the latest changes in 2c2eca6 this is testing well for me now. @ryelle so that change would be very close to pulling in `@woocommerce/block-library` from npm correct? If so I think this is looking great!

**Comment by claudiosanches** [comment]
@ryelle why not just download or clone? For what I can see, it's using npm just to download it. We could just download it using Grunt, so we can do it directly from GitHub with no worry about servers access, so anyone can update it, this could be nice because we can skip any versioning for now and always download the latest version, or if need to fix in stable versions, we could move it as a composer package, so will be easy to set the version and make it be installed as a dev dependency, then that grunt task should do almost the same, just need to update all paths.

**Comment by ryelle** [comment]
> so that change would be very close to pulling in `@woocommerce/block-library` from npm correct?

Exactly, this zip file is what npm would download if I published it now.

@claudiosanches We do want to version this when it's launched, it will make tracking changes between the feature plugin and WC core easier. The zip download here is a temporary "proof of concept", so while we could set up the `package.json` reference to a github URL, we want the fixed versioning (which we're already doing with the plugin versions for the wp.org repo). We'll be using npm much like you're suggesting composer, but we're already publishing JS packages to npm. Plus, this follows the WordPress core method of publishing [`@wordpress/block-library` to npm.](http://npmjs.com/package/@wordpress/block-library)

**Comment by timmyc** [comment]
@ryelle what are the next steps for moving forward with this? Do we need to get  `@woocommerce/block-library` published first?

**Comment by claudiosanches** [comment]
@ryelle I'm suggesting composer because composer have some support for WP plugins too. Or if this is really intend to be a NPM package, we probably can start publishing it now. Or grabbing the URL from wp.org.

**Comment by ryelle** [comment]
>  what are the next steps for moving forward with this?

The point was to get a check of whether this approach would work, and it seems we're there, so yep – I'll push a `2.0.0-dev` as a prerelease to npm, and then update this PR to pull from the package npm. Then this PR can be moved out of "draft" stage and ready to review+merge.

Once we've wrapped the 2.0 milestone in the feature plugin, we'll release a `2.0.0-beta`, make another PR to update woo core to use that version, and then follow with PRs for RC & release after that.

So that said, I can publish `@woocommere/block-library` today, and keep this moving 🙂

**Comment by ryelle** [comment]
I've published [@woocommerce/block-library 2.0.0-alpha](https://www.npmjs.com/package/@woocommerce/block-library), so this branch has been updated to pull from npm properly now. I've updated the instructions above to reflect the final way to test, which will match the process going forward.

**Comment by timmyc** [comment]
Testing out flawlessly for me now:

```
➜  woocommerce git:(try/block-library-package) ✗ grunt blocks
Running "clean:blocks" (clean) task
>> 3 paths cleaned.

Running "copy:js" (copy) task
Copied 9 files

Running "copy:css" (copy) task
Copied 2 files

Running "copy:php" (copy) task
Copied 2 files

Done.
```

**Review by timmyc — APPROVED** [review]
With the published npm package this is working wonderfully. I say we merge this to the feature branch to get all things ready for a final code review.

**Pull request merged**